elasticdl/python/ps/servicer.py

```diff
@@ -25,8 +26,31 @@ def __init__(
         self._lock = threading.Lock()
 
     def pull_variable(self, request, _):
-        # TODO: implement this RPC service
-        return elasticdl_pb2.PullVariableResponse()
+        """
+        Response with all non-embedding parameters if initialized.
+        """
+        res = elasticdl_pb2.PullVariableResponse()
+        if not self._parameters.init_status:
+            res.model_init_status = False
+            return res
+
+        # Only sync-SGD needs lock
+        # TODO: use a read-write lock to support multiple concurrent reads
+        if not self._use_async:
+            self._lock.acquire()
+        res.model.version = self._parameters.version
+        for name in self._parameters.non_embedding_params:
+            tensor = res.model.param.add()
+            tensor.name = name
+            var = self._parameters.non_embedding_params[name]
```

Comment:
How about using `for name, var in self._parameters.non_embedding_params.items()`. Through this way, it does not need to query the python dictionary and hence it is faster.

Reply:
Fixed.